File: Design-Patterns-Using-Ruby/Factory/Factory.rb
# Product interface
class Animal
    def speak
        raise NotImplementedError, 'Subclasses must implement the speak method'
    end
end

# Concrete Product classes
class Dog < Animal
    def speak
        puts 'Woof Woof!'
    end
end

class Cat < Animal
    def speak
        puts 'Meow Meow!'
    end
end

# Creator abstract class
class AnimalFactory
    def create_animal
        raise NotImplementedError, 'Subclasses must implement the create_animal method'
    end
end

# Concrete Creator classes
class DogFactory < AnimalFactory
    def create_animal
        Dog.new
    end
end

class CatFactory < AnimalFactory
    def create_animal
        Cat.new
    end
end
  
# Client code
dog_factory = DogFactory.new
dog = dog_factory.create_animal
dog.speak

cat_factory = CatFactory.new
cat = cat_factory.create_animal
cat.speak

=begin 
Woof Woof!
Meow Meow!
=end

File: Design-Patterns-Using-Kotlin/Factory/Factory.kt
// Animal interface
interface Animal {
    fun speak()
}

// Concrete Animal classes
class Dog : Animal {
    override fun speak() {
        println("Woof!")
    }
}

// Concrete Animal classes
class Cat : Animal {
    override fun speak() {
        println("Meow!")
    }
}

// AnimalFactory abstract class
abstract class AnimalFactory {
    abstract fun createAnimal(): Animal
}

// Concrete DogFactory classes
class DogFactory : AnimalFactory() {
    override fun createAnimal(): Animal {
        return Dog()
    }
}

// Concrete CatFactory classes
class CatFactory : AnimalFactory() {
    override fun createAnimal(): Animal {
        return Cat()
    }
}

// Client code
fun main() {
    val dogFactory: AnimalFactory = DogFactory()
    val dog = dogFactory.createAnimal()
    dog.speak()

    val catFactory: AnimalFactory = CatFactory()
    val cat = catFactory.createAnimal()
    cat.speak()
}

/*
Woof!
Meow!
*/

